pyomo/contrib/cp/sequence_var.py

```diff
@@ -0,0 +1,151 @@
+#  ___________________________________________________________________________
+
+import logging
+
+from pyomo.common.log import is_debug_set
+from pyomo.common.modeling import NOTSET
+from pyomo.contrib.cp import IntervalVar
+from pyomo.core import ModelComponentFactory
+from pyomo.core.base.component import ActiveComponentData
+from pyomo.core.base.global_set import UnindexedComponent_index
+from pyomo.core.base.indexed_component import ActiveIndexedComponent
+from pyomo.core.base.initializer import Initializer
+
+import sys
+from weakref import ref as weakref_ref
+
+logger = logging.getLogger(__name__)
+
+
+class SequenceVarData(ActiveComponentData):
+    """This class defines the abstract interface for a single sequence variable."""
+
+    __slots__ = ('interval_vars',)
+
+    def __init__(self, component=None):
+        # in-lining ActiveComponentData and ComponentData constructors, as is
+        # traditional:
+        self._component = weakref_ref(component) if (component is not None) else None
+        self._index = NOTSET
+        self._active = True
+
+        # This thing is really just an ordered set of interval vars that we can
+        # write constraints over.
+        self.interval_vars = []
+
+    def set_value(self, expr):
+        # We'll demand expr be a list for now--it needs to be ordered so this
+        # doesn't seem like too much to ask
+        if not hasattr(expr, '__iter__'):
+            raise ValueError(
+                "'expr' for SequenceVar must be a list of IntervalVars. "
+                "Encountered type '%s' constructing '%s'" % (type(expr), self.name)
+            )
+        for v in expr:
+            if not hasattr(v, 'ctype') or v.ctype is not IntervalVar:
+                raise ValueError(
+                    "The SequenceVar 'expr' argument must be a list of "
+                    "IntervalVars. The 'expr' for SequenceVar '%s' included "
+                    "an object of type '%s'" % (self.name, type(v))
+                )
+            self.interval_vars.append(v)
+
+
+@ModelComponentFactory.register("Sequences of IntervalVars")
+class SequenceVar(ActiveIndexedComponent):
+    _ComponentDataClass = SequenceVarData
+
+    def __new__(cls, *args, **kwds):
+        if cls != SequenceVar:
+            return super(SequenceVar, cls).__new__(cls)
+        if args == ():
+            return ScalarSequenceVar.__new__(ScalarSequenceVar)
+        else:
+            return IndexedSequenceVar.__new__(IndexedSequenceVar)
+
+    def __init__(self, *args, **kwargs):
+        self._init_rule = Initializer(kwargs.pop('rule', None))
+        self._init_expr = kwargs.pop('expr', None)
+        kwargs.setdefault('ctype', SequenceVar)
+        super(SequenceVar, self).__init__(*args, **kwargs)
+
+        if self._init_expr is not None and self._init_rule is not None:
+            raise ValueError(
+                "Cannot specify both rule= and expr= for SequenceVar %s" % (self.name,)
+            )
```

Comment:
There is a method (`self._pop_from_kwargs()` that can combine this logic and the two pop() above.  See `constraint.py:660`